blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1066 @@
+    api: Arc<RwLock<T>>,
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    cur_tipset: Arc<RwLock<Tipset>>,
+    revert: Vec<Tipset>,
+    apply: Vec<Tipset>,
+) -> Result<(), Error>
+where
+    T: Provider,
+{
+    let mut rmsgs: HashMap<Address, HashMap<u64, SignedMessage>> = HashMap::new();
+    for ts in revert {
+        let api_locked = api.write().await;
+        let pts = api_locked.load_tipset(ts.parents())?;
+        drop(api_locked);
+
+        let mut msgs: Vec<SignedMessage> = Vec::new();
+        for block in ts.blocks() {
+            let api_locked = api.read().await;
+            let (umsg, mut smsgs) = api_locked.messages_for_block(&block)?;
+            drop(api_locked);
+            msgs.append(smsgs.as_mut());
+            for msg in umsg {
+                let bls_sig_cache = bls_sig_cache.clone();
+                let smsg = recover_sig(bls_sig_cache, msg).await?;
+                msgs.push(smsg)
+            }
+        }
+        let parent = pts.clone();
+        let mut cur_ts_locked = cur_tipset.write().await;
+        *cur_ts_locked = parent;
+        drop(cur_ts_locked);
+
+        for msg in msgs {
+            add(msg, rmsgs.borrow_mut());
+        }
+    }
+
+    for ts in apply {
+        for b in ts.blocks() {
+            let api_locked = api.read().await;
+            let (msgs, smsgs) = api_locked.messages_for_block(b)?;
+            drop(api_locked);
+
+            for msg in smsgs {
+                rm(
+                    msg.from(),
+                    pending.clone(),
+                    msg.sequence(),
+                    rmsgs.borrow_mut(),
+                )
+                .await;
+            }
+            for msg in msgs {
+                rm(
+                    msg.from(),
+                    pending.clone(),
+                    msg.sequence(),
+                    rmsgs.borrow_mut(),
+                )
+                .await;
+            }
+        }
+        let mut cur_ts_locked = cur_tipset.write().await;
+        *cur_ts_locked = ts;
+        drop(cur_ts_locked);
+    }
+    for (_, hm) in rmsgs {
+        for (_, msg) in hm {
+            let api_locked = api.clone();
+            let pending = pending.clone();
+            let bls_sig_cache = bls_sig_cache.clone();
+            add_helper(api_locked, bls_sig_cache, pending, msg.clone())
+                .await
+                .ok();
+        }
+    }
+    Ok(())
+}
+
+/// This is a helper method for head_change. This method will remove a sequence for a from address
+/// from the rmsgs hashmap. Also remove the from address and sequence from the mmessagepool.
+async fn rm(
+    from: &Address,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    sequence: u64,
+    rmsgs: &mut HashMap<Address, HashMap<u64, SignedMessage>>,
+) {
+    let s = rmsgs.get_mut(from);
+    if s.is_none() {
+        remove(from, pending.clone(), sequence).await.ok();
```

Comment:
can you not just put `.ok` on these? you're just ignoring the error, when it would otherwise be handled